go-controller/pkg/factory/factory.go

```diff
@@ -269,19 +265,19 @@ func (i *informer) newFederatedHandler() cache.ResourceEventHandlerFuncs {
 	}
 }
 
-func (i *informer) shutdown() {
+func (i *informer) removeAllHandlers() {
 	i.Lock()
 	defer i.Unlock()
-
 	for _, handler := range i.handlers {
 		_ = i.removeHandler(handler)
 	}
+}
 
-	// Close all event channels for queued informers
-	for idx := range i.events {
-		close(i.events[idx])
-		i.events[idx] = nil
-	}
+func (i *informer) shutdown() {
+	i.removeAllHandlers()
+
+	// Wait for all event processors to finish
+	i.shutdownWg.Wait()
```

Comment:
comment at L279 is only  valid for Queued Informers and not for others as it is going to immediately return.